source/RoslynAnalyzers/MessageContractAnalyzers.cs

```diff
@@ -0,0 +1,302 @@
+                        }
+
+                        break;
+                    case MessageType.Request requestType:
+                        if (RequestTypes_MustBeNamedCorrectly(context, typeDec))
+                        {
+                            // the expected response name depends on the request name; so only run this check if the RequestName was good
+                            RequestTypes_MustHaveCorrectlyNamedResponseTypes(context, typeDec, requestType.NameSyntax);
+                        }
+
+                        break;
+                }
+                // no specific checks for responses at this point
+            }
+            else if (apiSurfaceType is ApiSurfaceType.Event)
+            {
+                EventTypes_MustBeNamedCorrectly(context, typeDec);
+            }
+            // no specific checks for resources at this point
+        }
+
+        static bool CheckMessageTypeProperties(SyntaxNodeAnalysisContext context, TypeDeclarationSyntax typeDec, SpecialTypeDeclarations cachedTypes)
+        {
+            var result = true;
+            foreach (var propDec in typeDec.DescendantNodes().OfType<PropertyDeclarationSyntax>())
+            {
+                // only validate public properties
+                if (!propDec.Modifiers.Any(m => m.IsKind(SyntaxKind.PublicKeyword))) continue;
+
+                // we have a number of places where we treat collection types differently; front-load that info
+                var isCollectionType = IsCollectionType(context, propDec, cachedTypes);
+                var required = GetRequiredState(propDec);
+
+                // if anything returns false, propagate false outwards
+                result &= PropertiesOnMessageTypes_MustBeMutable(context, propDec);
+                result &= RequiredPropertiesOnMessageTypes_MustNotBeNullable(context, propDec, required);
+                result &= OptionalPropertiesOnMessageTypes_ExceptForCollections_MustBeNullable(context,
+                    propDec,
+                    required,
+                    isCollectionType,
+                    cachedTypes);
+                result &= MessageTypes_MustInstantiateCollections(context, propDec, required, isCollectionType);
+                result &= PropertiesOnMessageTypes_MustHaveAtLeastOneValidationAttribute(context, propDec, required);
+                result &= SpaceIdPropertiesOnMessageTypes_MustBeOfTypeSpaceId(context, propDec, cachedTypes);
+                result &= IdPropertiesOnMessageTypes_MustBeACaseInsensitiveStringTinyType(context, propDec, cachedTypes);
+            }
+
+            return result;
+        }
+
+        // ----- helpers --------------- 
+
+        // Don't think of this as an OOP class hierarchy; think of it as an Enum with associated values.
+        //
+        // it bugs me that these aren't structs because our syntax parser is now incurring allocations.
```

Comment:
The analyzer is no doubt allocating some strings and lambda's and these records don't matter... But still, it bugs me anyway :smile: Old C++ habits die hard

Reply:
Nice approach - i like it (well, except for the allocations 😄 ).
I can definitely see your C++ heritage coming through ;)